### DISCUSSION TITLE
Is there a way to merge shapes on a certain layer within the cell

### DISCUSSION THREAD

**Discussion by JinWallner** [gdsfactory/gdsfactory#933] [Q&A]
I am trying the SiEPIC/Ansys netlisting tool. It works well for most components but the bend. The DevRec marker generated by cladding is broken into four pieces and causes errors in component extraction. If I use bbox for DevRec, it works. But now the DevRec marker is big and often overlaps with each other if I have a bundle. Is there a way to merge the DevRec shapes generated by cladding?

Thanks,
Jin

**Comment by joamatab** [comment]
Hi Jin 

Where can we find the code to reproduce the issue?

**Comment by JinWallner** [comment]
Hi Joaquin,

Please check layer 68/0 in the bends generated by the code below.

Thanks,

Jin

c = gf.Component("Test")

c1 = c << gf.components.mmi2x2()
c2 = c << gf.components.mmi2x2()
c2.move((100, 40))
routes = gf.routing.get_bundle(
[c1.ports["o2"], c1.ports["o1"]],
[c2.ports["o1"], c2.ports["o2"]],
radius=5,
cross_section='strip'
)
for route in routes:
c.add(route.references)
c.show(show_ports=True)

**Reply by JinWallner** [reply]
I got it. It's the default npoints=720 that reached the GDS limit? If I reduce the points to 100, then the bend will be one polygon. Can the padding layer have a different number of points from the core layer?

**Reply by tvt173** [reply]
I would classify that as a SiEpic bug to be honest. This sounds like a fracturing issue. It is common for polygons with many vertices to get fractured in order to stay compatible with all downstream tools. A proper application should merge the geometries before trying to use them, in case they have been fractured like this. Should be fixed in SiEpic.

FYI default limit in gdstk for a polygon to get fractured is 199 points.
https://heitzmann.github.io/gdstk/library/gdstk.GdsWriter.html

**Comment by JinWallner** [comment]
I have another question related with merge. In the example below, t1 has port snapping issues while t2 seems ok. But will the t2 method always work? Usually, we will merge the big piece with those small pieces before any rotation operation. Is there any mechanism in gdsfactory to avoid potential port snapping issues in this kind of case or there is a safe way to do it?
Thanks,
Jin

c = gf.Component("Test")
@gf.cell
def fork()->gf.Component:
    c = gf.Component()
    s1 = gf.Section(width=0.4, offset=0.7, layer=(1,0), name='top')
    s2 = gf.Section(width=0.4, offset=-0.7, layer=(1,0), name='bot')
    X1 = gf.CrossSection(
        width=0.4,
        offset=0,
        layer=(1,0),
        name='cen',
        sections=[s1, s2],
    )

    s_1 = c << gf.components.straight(length=10,cross_section=X1)
    c.add_port(name="o2",port=s_1.ports['o2'])
    c.add_port(name="o1",port=s_1.ports['o1'])
    return c

@gf.cell
def rotate_test()->gf.Component:
    c = gf.Component()
    s_1 = c << gf.components.straight(length=10, layer=(1,0),width=2.4)
    s_2 = c << fork()
    s_2.rotate(2.5)
    s_1.connect('o2', s_2.ports['o1'])
    return c

@gf.cell
def rotate_test2()->gf.Component:
    c = gf.Component()
    s_1 = c << gf.components.straight(length=10, layer=(1,0),width=2.4)
    s_2 = c << fork()
    s_1.connect('o2', s_2.ports['o1'])
    return c.flatten()

t1 = c << rotate_test()
t2 = c << rotate_test2()
t2.move((0, 30))
t2.rotate(2.5)
c.show()

**Reply by tvt173** [reply]
hi @JinWallner , if I understand your question, you're talking about cases where you have non-90-degree rotations and the port geometries are not snapping properly, right? Similar to this issue?
https://gdsfactory.github.io/gdsfactory/notebooks/08_pdk.html#PDK-decorator

I recommend setting `flatten_invalid_refs` as your PDK's `default_decorator`. Then you should no longer see snapping errors at your ports (make sure your top level is wrapped with `@cell` to see this in action, otherwise the decorator will not take effect).

**Reply by JinWallner** [reply]
Hi @tvt173, thank you for the suggestion. The flatten_invalid_refs helps a lot. But in some cases, I need to abut 2 ports of different widths and I still see snapping issues at non-90-degree rotations. To make sure there are no gaps, additional points will need to be added to the wider port. 
I did figure out how to merge the shapes though. The union by layer works well.

**Reply by tvt173** [reply]
this is actually an impossible problem to solve, rooted in the GDS format. since vertices snap to integer grid coordinates, you literally cannot guarantee that diagonal lines of different widths will exactly overlap. it is a shortcoming you'll have to live with when working with GDS files. but like you say, you can add additional vertices to the wider face if necessary

**Comment by joamatab** [comment]
Hi Jin, the easiest solution for that was to increase the max number of polygons when saving the GDS file 

We did that 2 months ago as the default to solve the SiEPIC issues for the edx course

**Comment by JinWallner** [comment]
Hi Joaquin,

I am concerned that if the route is very long and complicated, the number of points might exceed the maximum number of points in the polygon.  A more robust way will be supporting simplifying/reducing points for cover layers. But it might be a more complicated solution.
Thanks,
Jin

**Reply by tvt173** [reply]
this can be done by merging the shapes with an OR operation and simplifying, i.e. with shapely
https://shapely.readthedocs.io/en/stable/reference/shapely.simplify.html?highlight=simplify

but i still consider this a bug on the side of SiEpic. they should be merging the polygons in their check.

**Reply by JinWallner** [reply]
Thanks. I did try to merge the shape in Klayout manually before running the SiEPIC tools and it worked.
